Implement div(E) diagnostics for spectral case

EZoni opened the pull request:
**Goal** Implement a spectral version of the `div(E)` diagnostics. When the PSATD solver is used, `div(E)` is now computed as the inverse Fourier transform of `i*k*FT(E)`, where `FT(E)` is the Fourier transform of `E`, instead of using finite differences.

The base class `SpectralBaseAlgorithm` has a new public member function named `ComputeSpectralDivE` that performs this calculation. A public interface has been added to the class `SpectralSolver` in order to call `ComputeSpectralDivE` directly from objects of class `SpectralSolver` (similar to the public interfaces `ForwardTransform`, `BackwardTransform` and `pushSpectralFields`).

Summary of main changes:

- **Source/FieldSolver/SpectralSolver/SpectralAlgorithms/SpectralBaseAlgorithm.H**: declaration of the new member function `ComputeSpectralDivE`;
- **Source/FieldSolver/SpectralSolver/SpectralAlgorithms/SpectralBaseAlgorithm.cpp** (new file): implementation of the new member function `ComputeSpectralDivE`;
- **Source/FieldSolver/SpectralSolver/SpectralSolver.H**: declaration and implementation of a public interface for the new member function `ComputeSpectralDivE` (the public interface is a member function of the class `SpectralSolver` and it is named again `ComputeSpectralDivE`);
- **Source/Diagnostics/FieldIO.cpp**: add call to the new function from the object `spectral_solver_fp` (object of class `SpectralSolver` defined on the fine patch), only when the PSATD solver is used.

**Test** I attach 4 figures showing the differences between the old diagnostic based on finite differences and the new spectral diagnostic for the test `pml_x_psatd`, where such differences are significant. The figures show 2D plots (on the (x,y) plane) of `rho/epsilon_0`, `Ez`, and `div(E)` computed with the two alternative methods after 100 iterations.

Consistently with the plots, the relative L-infinity spatial norm of `rho/epsilon_0-div(E)` (namely: `max_(x,y)(|rho/epsilon_0-div(E)|)/max_(x,y)(|rho/epsilon_0|)`) is ~ 11.85 for the finite differences case and ~ 0.01269 for the spectral case.

[pml_x_psatd_rho.pdf](https://github.com/ECP-WarpX/WarpX/files/4207444/pml_x_psatd_rho.pdf)
[pml_x_psatd_Ez.pdf](https://github.com/ECP-WarpX/WarpX/files/4207445/pml_x_psatd_Ez.pdf)
[pml_x_psatd_divE_fd.pdf](https://github.com/ECP-WarpX/WarpX/files/4207446/pml_x_psatd_divE_fd.pdf)
[pml_x_psatd_divE_sp.pdf](https://github.com/ECP-WarpX/WarpX/files/4207447/pml_x_psatd_divE_sp.pdf)

**Warning** The test `pml_x_psatd` now checks also the output of `div(E)` vs `rho/epsilon_0`: the corresponding benchmark on Battra needs to be added.

dpgrote commented:
When I implemented the k-space filtering in the RZ spectral solver, I took a different approach in how to get the k-vectors. What I did is to save the k_space object as a member of the class (instead of having it only as a temporary in the init routine). This avoids having to recalculate the k-vectors (which is somewhat expensive in the RZ-solver) and also avoids having to save the various other input parameters, such as the dm object and orders.

I also perform the actual multiplications in the RZ version of the SpectralFieldData class. It seems more appropriate since that is where the fields object is saved and where other loops over the array are defined.

EZoni commented:
> When I implemented the k-space filtering in the RZ spectral solver, I took a different approach in how to get the k-vectors. What I did is to save the k_space object as a member of the class (instead of having it only as a temporary in the init routine). This avoids having to recalculate the k-vectors (which is somewhat expensive in the RZ-solver) and also avoids having to save the various other input parameters, such as the dm object and orders.
> 
> I also perform the actual multiplications in the RZ version of the SpectralFieldData class. It seems more appropriate since that is where the fields object is saved and where other loops over the array are defined.

Thanks for the feedback. OK, I can move the new member function from the class `SpectralSolver` to the class `SpectralFieldData` and save the `SpectralKSpace` object `k_space` and the k-vectors as members of that class.

RemiLehe commented:
Actually, the `PsatdAlgorithm` base class already stores the `k` vectors. Maybe it would make sense to implement the computation of `divE` as a member function of this class. 
@EZoni We could maybe discuss in person today.

EZoni commented:
> Actually, the `PsatdAlgorithm` base class already stores the `k` vectors. Maybe it would make sense to implement the computation of `divE` as a member function of this class.
> @EZoni We could maybe discuss in person today.

I moved the computation of `div(E)` to the base class `SpectralBaseAlgorithm` and added a public interface in the class `SpectralSolver` in order to call the new function from objects of this latter class. This is explained a bit in more detail in the updated PR description. As discussed with @RemiLehe, I also simplified a bit my previous implementation by removing unnecessary pointers.

EZoni commented:
> This looks good overall. I have one concern. Adding divE to the SpectralFieldIndex adds more memory to the fields array in SpectralFieldData, memory that will only be used when divE is calculated. This memory will presumably be allocated on the GPU but will be mostly sitting around unused. Can one of the other indices, such as Bx, be reused instead of adding a new one?

Yes, @RemiLehe also mentioned this point (offline, last week) and I think he would like to change this behavior in general in a separate PR, as far as I understood.

dpgrote commented:
Ok. Why a separate PR?

EZoni commented:
> Ok. Why a separate PR?

I think @RemiLehe wants to define `SpectralFieldIndex` in a different way, without using an enumeration inside a structure, in order to avoid the memory issue that you mention. I think he has something in mind already, but we have not discussed this thoroughly. This is conceptually quite independent of the fact that we are currently adding a new field within this PR. Hence, as far as I understood, he would prefer to do this in a separate small PR. Maybe we can wait for @RemiLehe to follow up on this. I would definitely be willing to implement the changes that he has in mind, either in this PR or in a separate one.

dpgrote commented:
If that is a small PR, then maybe this one can wait until that one is merged. I want to see what the suggested change is since this will affect the RZ spectral solver (since it is using the same SpectralFieldIndex).

RemiLehe commented on Source/FieldSolver/SpectralSolver/SpectralFieldData.H:
```diff
@@ -19,7 +19,7 @@ using SpectralField = amrex::FabArray< amrex::BaseFab <Complex> >;
 /** Index for the regular fields, when stored in spectral space */
 struct SpectralFieldIndex {
-  enum { Ex=0, Ey, Ez, Bx, By, Bz, Jx, Jy, Jz, rho_old, rho_new, n_fields };
+  enum { Ex=0, Ey, Ez, Bx, By, Bz, Jx, Jy, Jz, rho_old, rho_new, divE, n_fields };
```
```suggestion
  enum { Ex=0, Ey, Ez, Bx, By, Bz, Jx, Jy, Jz, rho_old, rho_new, n_fields
              divE=3 }; // divE reuses the memory slot for Bx, since Bx is not used when computing divE
```

RemiLehe commented:
What I had in mind was to replace the multi-component MultiFab that contains the all fields in `k` space by an array of pointers to single-component MultiFab that each contain a single field. This way, we would have more freedom to allocate individual fields as needed.

However, I think that this would need some more time and discussion.

For now, in order to finalize this PR, I agree with Dave's suggestion to reuse the index for `Bx`. I added a corresponding code suggestion (see above).
@EZoni If you agree with the above suggestion, could you commit that change in this PR?
